File: ingestion/python_api/EtLT/load/db_helpers.py
# =============================================================================
# Purpose: Shared MySQL utilities for the finance pipeline.
#
#   Provides the database connection factory and FK-resolution helpers used
#   by all loader modules (load/plans.py, load/plan_items.py,
#   load/transactions.py).
#
#   Public API:
#     get_connection()        – pymysql.Connection
#     _build_lookup_maps()    – name→id dicts for accounts/categories/etc.
#     _resolve()              – case-insensitive name → id with warning
#     _parse_amount()         – string → int (VND, no decimals)
# =============================================================================

import os

import pymysql
import pymysql.cursors


# ─── Connection ───────────────────────────────────────────────────────────────

def get_connection() -> pymysql.Connection:
    return pymysql.connect(
        host     = os.getenv("MYSQL_HOST",     "mysql"),
        user     = os.getenv("MYSQL_USER",     "root"),
        password = os.getenv("MYSQL_PASSWORD", ""),
        database = os.getenv("MYSQL_DATABASE", "finance_db"),
        charset  = "utf8mb4",
        cursorclass = pymysql.cursors.DictCursor,
    )


# ─── Lookup helpers ───────────────────────────────────────────────────────────

def _build_lookup_maps(cursor) -> dict:
    """
    Load reference tables once per call and return name→id dicts.
    Keys in returned dict: 'account', 'category', 'subcategory', 'plan'.
    Plan map key: (year: int, month: int) → plan_id.
    """
    cursor.execute("SELECT account_id, account_name FROM accounts")
    accounts = {r["account_name"].strip().lower(): r["account_id"]
                for r in cursor.fetchall()}

    cursor.execute("SELECT category_id, category_name FROM categories")
    categories = {r["category_name"].strip().lower(): r["category_id"]
                  for r in cursor.fetchall()}

    cursor.execute("SELECT subcategory_id, subcategory_name FROM subcategories")
    subcategories = {r["subcategory_name"].strip().lower(): r["subcategory_id"]
                     for r in cursor.fetchall()}

    cursor.execute("SELECT plan_id, plan_year, plan_month FROM monthly_plans")
    plans = {(r["plan_year"], r["plan_month"]): r["plan_id"]
             for r in cursor.fetchall()}

    return {
        "account":     accounts,
        "category":    categories,
        "subcategory": subcategories,
        "plan":        plans,
    }


def _resolve(lookup_map: dict, key: str, context: str):
    """Case-insensitive name → id.  Returns None and prints a warning if missing."""
    if not key or str(key).strip() == "":
        return None
    result = lookup_map.get(str(key).strip().lower())
    if result is None:
        print(f"[db_service] WARNING: cannot resolve {context}='{key}' — will insert NULL")
    return result


def _parse_amount(value) -> int:
    """Strip commas/whitespace → int (VND has no decimals)."""
    if value is None or str(value).strip() == "":
        return 0
    cleaned = str(value).replace(",", "").replace(".", "").strip()
    try:
        return int(float(cleaned))
    except (ValueError, TypeError):
        return 0

File: ingestion/nodejs_fallback/src/EtLT/load/DbHelpers.ts
// =============================================================================
// File: EtLT/load/DbHelpers.ts
// Purpose: Shared MySQL utilities — mirrors Python's EtLT/load/db_helpers.py.
//   Provides connection factory and FK-resolution helpers used by all
//   loader modules (plans.ts, plan_items.ts, transactions.ts).
// =============================================================================

import mysql from "mysql2/promise";

export interface LookupMaps {
  account:     Record<string, number>;
  category:    Record<string, number>;
  subcategory: Record<string, number>;
  plan:        Record<string, number>;
}

export function getConnection(): Promise<mysql.Connection> {
  return mysql.createConnection({
    host:     process.env.MYSQL_HOST     ?? "mysql",
    user:     process.env.MYSQL_USER     ?? "finance_user",
    password: process.env.MYSQL_PASSWORD ?? "",
    database: process.env.MYSQL_DATABASE ?? "finance_db",
    charset:  "utf8mb4",
  });
}

export async function buildLookupMaps(conn: mysql.Connection): Promise<LookupMaps> {
  const [accounts]      = await conn.execute<any[]>("SELECT account_id, account_name FROM accounts");
  const [categories]    = await conn.execute<any[]>("SELECT category_id, category_name FROM categories");
  const [subcategories] = await conn.execute<any[]>("SELECT subcategory_id, subcategory_name FROM subcategories");
  const [plans]         = await conn.execute<any[]>("SELECT plan_id, plan_year, plan_month FROM monthly_plans");

  return {
    account:     Object.fromEntries(accounts.map(     (r: any) => [r.account_name.trim().toLowerCase(),     r.account_id])),
    category:    Object.fromEntries(categories.map(   (r: any) => [r.category_name.trim().toLowerCase(),    r.category_id])),
    subcategory: Object.fromEntries(subcategories.map((r: any) => [r.subcategory_name.trim().toLowerCase(), r.subcategory_id])),
    plan:        Object.fromEntries(plans.map(        (r: any) => [`${r.plan_year}-${String(r.plan_month).padStart(2, "0")}`, r.plan_id])),
  };
}

/** Case-insensitive name → id. Returns null and logs a warning if missing. */
export function resolve(
  map: Record<string, number>,
  key: string,
  context: string
): number | null {
  if (!key || key.trim() === "") return null;
  const result = map[key.trim().toLowerCase()];
  if (result === undefined) {
    console.warn(`[DbHelpers] WARNING: cannot resolve ${context}='${key}' — will insert NULL`);
    return null;
  }
  return result;
}

/** Strip commas/whitespace → integer (VND has no decimals). */
export function parseAmount(value: any): number {
  if (value == null || String(value).trim() === "") return 0;
  const cleaned = String(value).replace(/,/g, "").trim();
  const num = parseFloat(cleaned);
  return isNaN(num) ? 0 : Math.round(num);
}
